File: Ray Tracing next Week/util/week2/ray.py
# ray.py

from util.week1.vec3 import Vec3

# Point3 is just Vec3 (same as before)
Point3 = Vec3


class Ray:
    def __init__(self, origin, direction, time=0.0):
        """
        OLD:
            def __init__(self, origin, direction)

        NEW:
            Added 'time'

        PURPOSE:
            Stores when this ray exists.
            Needed for motion blur.
        """
        
        # Same as before
        self.orig = origin
        
        # Same as before
        self.dir = direction
        
        # NEW: store ray time
        self.tm = time

    def origin(self):
        """
        Same as before.
        Returns ray starting point.
        """
        return self.orig

    def direction(self):
        """
        Same as before.
        Returns ray direction.
        """
        return self.dir

    def time(self):
        """
        NEW function.

        Returns time of this ray.
        Used by moving objects.
        """
        return self.tm

    def at(self, t):
        """
        Same formula as before:

        P(t) = O + tD

        where:
            O = origin
            D = direction
            t = distance

        Returns point along ray.
        """
        return self.orig + t * self.dir

File: Ray Tracing next Week - Javascript/Ch2/ray.js
class Ray{
    constructor({origin,direction,time=0}){
      //console.log("Time :",time)
      this.A = origin
      this.B = direction
      this.tm = time
    }

     origin(a){this.A = a}

     origin(){return this.A}
     direction(){return this.B}
     time(){return this.tm}

    at(t){
        // Scales vector B by scalar t, then adds origin vector A
    const scaled_dir = Vec3.mul({ t: t, v: this.B });
    return Vec3.add(this.A, scaled_dir);
    
    }
}

window.Ray = Ray